fscoordinator stat caching

lordnull opened the pull request:
To avoid blocking the full sync coordinator when fetching stats, having a temporary process spawn up and do it on an interval for the coordinator instead.

lordnull committed:
Fixed compile and run-time errors.

lordnull committed:
Made refresh timer configurable.

Default can be configured as a macro, but run-time value is taken from
app config.

lordnull committed:
Added fullsync stat cache interval to config schema

cmeiklejohn commented:
This is a must for 2.0.

cmeiklejohn commented:
I'd really like for there to be a r_t that demonstrates the problem and can be used to validate this fix.

andrewjstone commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -27,6 +27,22 @@
 % next partition.
 -define(WAITING_TIMEOUT, 5000).
+% How often stats should be cached in milliseconds.
+-ifndef(STAT_REFRESH_INTERVAL).
```
should this be `DEFAULT_STAT_REFRESH_INTERVAL` ?

lordnull committed:
Corrected an ifndef check.

andrewjstone commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -27,6 +27,22 @@
+-endif.
+-endif.
+
+-record(stat_cache, {
```
Just a minor nitpick, but IMO records should have all fields typed. I also think this may generate a dialyzer warning in the future if we decide to turn on warnings for untyped records.

andrewjstone commented:
- [ ] code inspection
- [x] eunit + eqc
- [ ] replication2 riak_test
- [x] repl_fs_stat_caching riak_test
- [x] dialyzer
- [x] xref

andrewjstone commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -740,6 +780,52 @@ sort_partitions(In, N, Acc) ->
+
+schedule_stat_refresh(StatCache) ->
+    StatCache1 = maybe_cancel_timer(StatCache),
+    Timer = erlang:send_after(?DEFAULT_STAT_REFRESH_INTERVAL, self(), refresh_stats),
```
This should use `StatCache#stat_cache.refresh_interval` instead of `?DEFAULT_STAT_REFRESH_INTERVAL`

andrewjstone commented:
I'm seeing the following dialyzer warning:

``` erlang
riak_repl2_fscoordinator.erl:501: The call riak_repl2_fscoordinator:refresh_stats(StatCache::#stat_cache{worker::{_,_}}) will never return since it differs in the 1st argument from the success typing arguments: (#state{leader_node::atom(),leader_pid::atom(),stat_cache::#stat_cache{worker::'undefined' | {pid() | port(),reference()}}})
```

lordnull committed:
Fixed dialzer warnings.

Added types to the stat_cache record. Also resovled dialzyer warnings
concerning added types.

lordnull commented:
I'm working on a riak_test for this. Before the merge, it's likely best this be rebased.

lordnull commented:
There is now a riak_test for this: https://github.com/basho/riak_test/pull/546

lordnull commented:
I've rebased these changes in #554; r_t still passes for me, but it may be worth a look over just to be sure.

andrewjstone commented:
Thanks @lordnull I'll take a look a #554 now. Last night replication2 was failing on this branch, but that could just be due to needing this rebase. Hopefully I will +1 soon.

lordnull commented:
This has been handled in another pull, making this one pointless.
